tests/test_nhg.py

```diff
@@ -138,3 +139,187 @@ def test_route_nhg(self, dvs, testlog):
                 for v in fvs:
                     if v[0] == "SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_GROUP_ID":
                         assert v[1] == nhgid
+
+    def test_route_nhg_exhaust(self, dvs, testlog):
+        """
+        Test the situation of exhausting ECMP group, assume SAI_SWITCH_ATTR_NUMBER_OF_ECMP_GROUPS is 512
+
+        In order to achieve that, we will config
+            1. 9 ports
+            2. 512 routes with different nexthop group
+
+        See Also
+        --------
+        SwitchStateBase::set_number_of_ecmp_groups()
+        https://github.com/Azure/sonic-sairedis/blob/master/vslib/src/SwitchStateBase.cpp
+
+        """
+
+        # TODO: check ECMP 512
+
+        def port_name(i):
+            return "Ethernet" + str(i * 4)
+
+        def port_ip(i):
+            return "10.0.0." + str(i * 2)
+
+        def peer_ip(i):
+            return "10.0.0." + str(i * 2 + 1)
+
+        def port_ipprefix(i):
+            return port_ip(i) + "/31"
+
+        def port_mac(i):
+            return "00:00:00:00:00:0" + str(i)
+
+        def gen_ipprefix(r):
+            """ Construct route like 2.X.X.0/24 """
+            ip = ipaddress.IPv4Address(IP_INTEGER_BASE + r * 256)
+            ip = str(ip)
+            ipprefix = ip + "/24"
+            return ipprefix
+
+        def gen_nhg_fvs(binary):
+            nexthop = []
+            ifname = []
+            for i in range(MAX_PORT_COUNT):
+                if binary[i] == '1':
+                    nexthop.append(peer_ip(i))
+                    ifname.append(port_name(i))
+
+            nexthop = ','.join(nexthop)
+            ifname = ','.join(ifname)
+            fvs = swsscommon.FieldValuePairs([("nexthop", nexthop), ("ifname", ifname)])
+            return fvs
+
+        def asic_route_exists(keys, ipprefix):
+            for k in keys:
+                rt_key = json.loads(k)
+
+                if rt_key['dest'] == ipprefix:
+                    return k
+            else:
+                return None
+
+        def asic_route_nhg_fvs(k):
+            (status, fvs) = rtbl.get(k)
```

Comment:
There's a `getEntry` method in dvs database that will handle this. The main differences are that it doesn't return a status and the return type is a dict, so you'll need to update a few other parts of the code as well.